Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial/Basic.lean

```diff
@@ -0,0 +1,432 @@
+/-
+Copyright (c) 2023 David Kurniadi Angdinata. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Kurniadi Angdinata
+-/
+import Mathlib.AlgebraicGeometry.EllipticCurve.Affine
+import Mathlib.NumberTheory.EllipticDivisibilitySequence
+
+/-!
+# Division polynomials of Weierstrass curves
+
+This file defines certain polynomials associated to division polynomials of Weierstrass curves and
+computes their leading terms. These are defined in terms of the auxiliary sequences of normalised
+elliptic divisibility sequences defined in `Mathlib.NumberTheory.EllipticDivisibilitySequence`.
+
+## Mathematical background
+
+Let $W$ be a Weierstrass curve over a commutative ring $R$. The sequence of $n$-division polynomials
+$\psi_n \in R[X, Y]$ of $W$ is the normalised elliptic divisibility sequence with initial values
+ * $\psi_0 := 0$,
+ * $\psi_1 := 1$,
+ * $\psi_2 := 2Y + a_1X + a_3$,
+ * $\psi_3 := 3X^4 + b_2X^3 + 3b_4X ^ 2 + 3b_6X + b_8$, and
+ * $\psi_4 := \psi_2 \cdot
+    (2X^6 + b_2X^5 + 5b_4X^4 + 10b_6X^3 + 10b_8X^2 + (b_2b_8 - b_4b_6)X + (b_4b_8 - b_6^2))$.
+
+Furthermore, define the associated sequences $\phi_n, \omega_n \in R[X, Y]$ by
+ * $\phi_n := X\psi_n^2 - \psi_{n + 1}\psi_{n - 1}$, and
+ * $\omega_n := \tfrac{1}{2} \cdot (\psi_{2n} / \psi_n - \psi_n(a_1\phi_n + a_3\psi_n^2))$.
+
+Note that $\omega_n$ is always well-defined as a polynomial. As a start, it can be shown by
+induction that $\psi_n$ always divides $\psi_{2n}$, so that $\psi_{2n} / \psi_n$ is always
+well-defined as a polynomial, while division by 2 is well-defined when $R$ has characteristic
+different from 2. In general, it can be shown that 2 always divides the polynomial
+$\psi_{2n} / \psi_n - \psi_n(a_1\phi_n + a_3\psi_n^2)$ in the characteristic zero universal ring
+$\mathcal{R}[X, Y] := \mathbb{Z}[a_i][X, Y] / \langle W(a_i, X, Y)\rangle$ associated to $W$, where
+$W(a_i, X, Y)$ is the associated Weierstrass equation. Then $\omega_n$ can be equivalently defined
+as its image under the associated universal morphism $\mathcal{R}[X, Y] \to R[X, Y]$.
```

Comment:
My suggestion:
```suggestion
$\psi_{2n} / \psi_n - \psi_n(a_1\phi_n + a_3\psi_n^2)$ in the characteristic zero universal
polynomial ring $\mathcal{R}[X, Y] := \mathbb{Z}[a_i][X, Y]$,
where $a_i$ are indeterminates. Then $\omega_n$ can be equivalently defined
as its image under the associated universal morphism $\mathcal{R}[X, Y] \to R[X, Y]$.
```
It's not necessary to quotient by the Weierstrass polynomial here. I'll modify the end of this paragraph in my PR to say that I actually take a more direct approach to define $\omega_n$.